File: android/app/src/main/java/it/curzel/bitscape/controller/ControllerSettingsStorage.kt
package it.curzel.bitscape.controller

import android.content.Context
import android.content.SharedPreferences
import androidx.compose.ui.unit.Dp

class ControllerSettingsStorage(
    context: Context,
    screenWidthDp: Dp,
    screenHeightDp: Dp
) {
    private val sharedPreferences: SharedPreferences =
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

    private companion object {
        const val PREFS_NAME = "ControllerSettingsPrefs"
        const val KEY_DID_SET_DEFAULTS = "kControllerOffsetDidSetDefaults"
        const val KEY_PORTRAIT_X = "kControllerOffsetPortraitX"
        const val KEY_PORTRAIT_Y = "kControllerOffsetPortraitY"
        const val KEY_LANDSCAPE_X = "kControllerOffsetLandscapeX"
        const val KEY_LANDSCAPE_Y = "kControllerOffsetLandscapeY"
        const val KEY_UNKNOWN = "kControllerOffsetUnknown"
    }

    init {
        if (!didSetDefaults()) {
            loadDefaults(screenWidthDp, screenHeightDp)
            setDefaultsLoaded()
        }
    }

    fun store(offset: Float, axis: ControllerOffsetAxis, orientation: ControllerOrientation) {
        val key = getKey(axis, orientation)
        sharedPreferences.edit().putFloat(key, offset).apply()
    }

    fun offset(axis: ControllerOffsetAxis, orientation: ControllerOrientation): Float {
        val key = getKey(axis, orientation)
        return sharedPreferences.getFloat(key, 0f)
    }

    private fun getKey(axis: ControllerOffsetAxis, orientation: ControllerOrientation): String {
        return when (axis to orientation) {
            ControllerOffsetAxis.X to ControllerOrientation.PORTRAIT -> KEY_PORTRAIT_X
            ControllerOffsetAxis.Y to ControllerOrientation.PORTRAIT -> KEY_PORTRAIT_Y
            ControllerOffsetAxis.X to ControllerOrientation.LANDSCAPE -> KEY_LANDSCAPE_X
            ControllerOffsetAxis.Y to ControllerOrientation.LANDSCAPE -> KEY_LANDSCAPE_Y
            else -> KEY_UNKNOWN
        }
    }

    private fun didSetDefaults(): Boolean {
        return sharedPreferences.getBoolean(KEY_DID_SET_DEFAULTS, false)
    }

    private fun setDefaultsLoaded() {
        sharedPreferences.edit().putBoolean(KEY_DID_SET_DEFAULTS, true).apply()
    }

    private fun loadDefaults(screenWidth: Dp, screenHeight: Dp) {
        val portraitX = ((screenWidth / 2) - keyEmulatorViewSize).value
        store(portraitX, ControllerOffsetAxis.X, ControllerOrientation.PORTRAIT)
        val portraitY = ((screenHeight / 2) - keyEmulatorViewSize).value - 150f
        store(portraitY, ControllerOffsetAxis.Y, ControllerOrientation.PORTRAIT)
        val landscapeX = ((screenHeight / 2) - keyEmulatorViewSize).value - 40f
        store(landscapeX, ControllerOffsetAxis.X, ControllerOrientation.LANDSCAPE)
        val landscapeY = ((screenWidth / 2) - keyEmulatorViewSize).value - 100f
        store(landscapeY, ControllerOffsetAxis.Y, ControllerOrientation.LANDSCAPE)
    }
}

enum class ControllerOrientation {
    PORTRAIT,
    LANDSCAPE
}

enum class ControllerOffsetAxis {
    X,
    Y
}

File: ios/SneakBit/Controller/ControllerSettingsStorage.swift
import Foundation
import Schwifty

protocol ControllerSettingsStorage {
    func store(offset: CGFloat, axis: ControllerOffsetAxis, orientation: ControllerOrientation)
    func offset(axis: ControllerOffsetAxis, orientation: ControllerOrientation) -> CGFloat
}

enum ControllerOrientation {
    case portrait
    case landscape
}

enum ControllerOffsetAxis {
    case x
    case y
}

class ControllerSettingsStorageImpl: ControllerSettingsStorage {
    private var defaults: UserDefaults {
        UserDefaults.standard
    }
    
    init() {
        if !didSetDefaults() {
            loadDefaults()
            defaults.set(true, forKey: kControllerOffsetDidSetDefaults)
        }
    }
    
    func store(offset: CGFloat, axis: ControllerOffsetAxis, orientation: ControllerOrientation) {
        defaults.set(Float(offset), forKey: key(axis: axis, orientation: orientation))
    }
    
    func offset(axis: ControllerOffsetAxis, orientation: ControllerOrientation) -> CGFloat {
        CGFloat(defaults.float(forKey: key(axis: axis, orientation: orientation)))
    }
    
    private func key(axis: ControllerOffsetAxis, orientation: ControllerOrientation) -> String {
        switch (axis, orientation) {
        case (.x, .portrait): kControllerOffsetPortraitX
        case (.y, .portrait): kControllerOffsetPortraitY
        case (.x, .landscape): kControllerOffsetLandscapeX
        case (.y, .landscape): kControllerOffsetLandscapeY
        }
    }
    
    private func didSetDefaults() -> Bool {
        defaults.bool(forKey: kControllerOffsetDidSetDefaults)
    }
    
    private func loadDefaults() {
        let width = min(Screen.main?.bounds.width ?? 0, Screen.main?.bounds.height ?? 0)
        let height = max(Screen.main?.bounds.width ?? 0, Screen.main?.bounds.height ?? 0)
        
        store(offset: width / 2 - KeyEmulatorView.size.width, axis: .x, orientation: .portrait)
        store(offset: height / 2 - KeyEmulatorView.size.height - 150, axis: .y, orientation: .portrait)
        
        store(offset: height / 2 - KeyEmulatorView.size.width - 40, axis: .x, orientation: .landscape)
        store(offset: width / 2 - KeyEmulatorView.size.height - 100, axis: .y, orientation: .landscape)
    }
}

private let kControllerOffsetDidSetDefaults = "kControllerOffsetDidSetDefaults"
private let kControllerOffsetPortraitX = "kControllerOffsetPortraitX"
private let kControllerOffsetPortraitY = "kControllerOffsetPortraitY"
private let kControllerOffsetLandscapeX = "kControllerOffsetLandscapeX"
private let kControllerOffsetLandscapeY = "kControllerOffsetLandscapeY"
